Hw04 auth

Marta7799 opened the pull request.

Marta7799 committed:
step 1

Marta7799 committed:
next step: controllers/users

Marta7799 committed:
Merge pull request #9 from Marta7799/hw03-mongodb

Hw03 mongodb

Marta7799 committed:
Update .env.example

Marta7799 committed:
in progress

Marta7799 committed:
some changes

Marta7799 committed:
Update users.js

KamilMr commented on api/contacts.js:
```diff
@@ -0,0 +1,13 @@
+const express = require("express");
+const router = express.Router();
+const contactController = require("../controllers/contacts");
+const { auth } = require("../middlewares/auth");
```
```suggestion
const express = require("express");

const contactController = require("../controllers/contacts");
const { auth } = require("../middlewares/auth");

const router = express.Router();
```

KamilMr replied:
Warto ale nie trzeba :) grupować sobie importy i układać alfabetycznie. Najpierw importy główne, potem importy zewnętrzne a na końcu importy wewnętrzne. Dzięki temu patrząc na plik szybko można się rozpoznać jak wygląda struktura tego pliku.

KamilMr commented on app.js:
```diff
@@ -9,7 +9,8 @@ const formatsLogger = app.get("env") === "development" ? "dev" : "short";
 app.use(cors());
 app.use(express.json());
-app.use("/api/contacts", require("./api"));
+app.use("/api/contacts", require("./api/contacts"));
```
Zauważyłem że dodajesz auth wszędzie w contacts route. Można się zastanowić czy nie dodać tego lepiej tu. Piszę to bardziej żeby się zastanowić niekoniecznie zmieniać.

Marta7799 replied:
Czyli wystarczy w app.js dodać: app.use("/api/auth", require("./middlewares/auth")); zamiast w routes?

KamilMr replied:
```suggestion
app.use("/api/contacts", auth, require("./api/contacts"));
```
Takie coś sprawi że middleware działa na wszystkich ścieżkach contacts.
Ps. ja nie zawsze widzę komentarze tutaj, wiec jakbym nie pisał to napisz na slack/lms ze potrzebujesz żebym się odniósł .

Marta7799 replied:
ok, dlatego właśnie na LMS napisałam :)

KamilMr commented on controllers/users.js:
```diff
@@ -0,0 +1,132 @@
+const service = require("../service/users");
+const jwt = require("jsonwebtoken");
+const User = require("../service/schemas/user");
+require("dotenv").config();
```
Gdyby na przykład grupować importy tutaj wyglądałoby to mniej więcej tak: 
```suggestion
require("dotenv").config();
const jwt = require("jsonwebtoken");

const User = require("../service/schemas/user");
const service = require("../service/users");
const { userValidator } = require("./../utils/validators/validator");
```

KamilMr commented on controllers/users.js:
```diff
@@ -0,0 +1,132 @@
+
+  const payload = {
+    id: user.id,
+    email: user.email,
```
Zakładam że `id` jest unique. W związku z tym nie potrzebuję tam podawać jeszcze email. Imię mniej informacji w payload tym lepiej bo pamiętajmy że to można łatwo odczytać.

KamilMr commented on middlewares/auth.js:
```diff
@@ -0,0 +1,43 @@
+const passportJWT = require("passport-jwt");
+const User = require("../service/schemas/user");
+require("dotenv").config();
+const secret = process.env.SECRET;
```
```suggestion
const secret = process.env.SECRET;
const SECRET = process.env.SECRET;
```
Ta linijka jest w porządku, natomiast  jeśli do stałej przypiszemy wartość która się nie zmienia, to pisze się je `SECRET` .  Pozwala to szybciej zrozumieć że ona jest na swój sposób wyjątkowa, w tym wypadku jest konfiguracją przekazaną z .env.

KamilMr commented on middlewares/auth.js:
```diff
@@ -0,0 +1,43 @@
+
+// JWT Strategy
+passport.use(
+  new Strategy(params, function (payload, done) {
```
```suggestion
  new Strategy(params,  (payload, done) => {
```
Może warto użyć arrow function gdyż taki sposób pisania Widzę że preferujesz

Marta7799 replied:
wzięte z konspektu dlatego nie zmieniałam:)

Marta7799 replied:
ale już zmienione :)

KamilMr commented on service/schemas/user.js:
```diff
@@ -0,0 +1,47 @@
+      type: String,
+      default: null,
+    },
+    owner: {
+      type: Schema.Types.ObjectId,
+      ref: "user",
+    },
```
Ten klucz powinien być dodany do kontaktów. Bo ty jesteś użytkownikiem który ma kontakty. Ja również loguje się na twoją aplikację i też jestem użytkownikiem ze swoimi kontaktami. W obecnej implementacji jaką masz Ja mam twoje kontakty i ty masz moje. Ten klucz powinien być dodany do schema contacts i kiedy zapisujesz kontakty, zmieniasz je to zawsze od tego momentu powinnaś sprawdzić czy dany kontakt należy do ciebie jeżeli nie pomimo tego że ID się zgadza powinnaś zwrócić nie ma takiego kontaktu. 

Mam nadzieję że ma to sens. :)

Marta7799 replied:
acha, czyli ma być tylko w contacts?

KamilMr replied:
Tak

Marta7799 replied:
ok

KamilMr reviewed (changes requested):
Hej, mała poprawka którą trzeba jeszcze dodać w tym zadaniu. Opis w PR. 
Trzeba zaimplementować logikę w której obsługujesz tylko swoje kontakty.

Marta7799 commented:
Poprawki wprowadzone, dwa pytanka tylko

Marta7799 committed:
after changes

Marta7799 committed:
Update contacts.js

Marta7799 committed:
shortening path for validator; removing auth from api files

trying to add id:owner to contacts controller

Marta7799 committed:
change letter in controller/contacts

KamilMr commented on api/users.js:
```diff
@@ -0,0 +1,12 @@
+
+router.post("/signup", userController.register);
+router.post("/login", userController.login);
+router.get("/logout", userController.logout);
+router.get("/current", userController.current);
```
Te dwie ścieżki powinny być authorized. To znaczy wylogować może się użytkownik który istnieje albo jest zalogowany. Jeśli dodasz tutaj `auth` wówczas w tych kontrolerach nie będziesz musiała pobierać danych ponownie ponieważ one już zostaną pobrane przez middleware. 
Ma sens?

KamilMr commented on service/schemas/user.js:
```diff
@@ -0,0 +1,43 @@
+  this.token = token;
+};
+
+const User = mongoose.model("User", userSchema, "users");
```
A ten trzeci argument jest tutaj potrzebny?

KamilMr reviewed (changes requested):
Hej, 
Dobrze Ci dzie wygląda coraz lepiej. Na filmiku więcej szczegółów :) 
Pozdrawiam

Marta7799 committed:
updated api, controllers, service

added authorization to router in users.js
changed token in users-controller
changed service/contacts

Marta7799 committed:
Update user.js

KamilMr commented on service/contacts.js:
```diff
@@ -1,27 +1,100 @@
+    throw new Error("Not found. Invalid Id!");
+  }
+  try {
+    const selectedConctact = await Contact.findOne({ _id: contactId });
```
Może dodać jeszcze owner klucz i wtedy szukasz po _id i po owner :) Skróci ci to trochę kod.

KamilMr commented on service/contacts.js:
```diff
@@ -1,27 +1,100 @@
 
-const createContact = async ({ name, email, phone, favorite }) => {
-  return Contact.create({ name, email, phone, favorite });
+    await Contact.findByIdAndDelete(contactId, body);
```
A sprawdzałeś czy nie da się wysłać więcej filtrów? Bo z tego co kojarzę to można dodać tez obiekt też.

Marta7799 replied:
Nie rozumiem...?

KamilMr replied:
coś takiego: 
```
db.getCollection("movies").find({genres: "Short", cast: "A.C. Abadie"})
```
Są dwa filtry, można zrobić _id i owner. I wtedy baza danych Ci sprawdzi czy taki rekord istnieje. Nie musisz tego sama sprawdzać potem.

Marta7799 replied:
Tylko, że ja już to wcześniej zmieniłam.... chyba znów nie wszedł mi push zanim sprawdziłeś, dlatego nie wiedizałam o co chodzi:(

KamilMr commented on service/contacts.js:
```diff
@@ -1,27 +1,100 @@
+    }
+
+    const { favorite } = body;
+    await Contact.findByIdAndUpdate(contactId, { favorite }, { new: true });
```
Sprawdzałeś czy ta metoda nie zwraca jakiś wartości na przykład nowo stworzonego użytkownika?

KamilMr commented on service/contacts.js:
```diff
@@ -1,27 +1,100 @@
+  }
+};
+
+const updateContact = async (contactId, body, userId) => {
```
Ta funkcja powinna robić updte a na dole widze ze wywoluje metode `findByIdAndDelete`

Marta7799 replied:
pomyłka :(

KamilMr reviewed (changes requested):
Hej,
Zobacz sobie z tymi metodami, niektóre z nich Zamiast zmieniać to usuwają. 
Upewnij się więc postmanie czy na pewno to wszystko się tam zgadza.

Walidacja którą umieściłaś jest w porządku natomiast można to przyśpieszyć w ten sposób że kiedy pobierasz dane użytkownika z bazy danych to prosisz bazę danych aby oprócz ID poszukała ci po kluczu owner.

Powodzenia

Marta7799 committed:
changed service and authorization

changed delete to uptade in service/contacts.js and changed authorization in middlewares/auth.js

Marta7799 commented:
na postmanie mi w ogóle nie działa :(
